v8go.cc

```diff
@@ -1297,6 +1297,38 @@ void ObjectSet(ValuePtr ptr, const char* key, ValuePtr prop_val) {
   obj->Set(local_ctx, key_val, prop_val->ptr.Get(iso)).Check();
 }
 
+RtnStrings ObjectGetPropertyNames(ValuePtr ptr) {
+  LOCAL_OBJECT(ptr);
+  RtnStrings rtn = {};
+
+  Local<Array> names;
+  if (!obj->GetPropertyNames(local_ctx).ToLocal(&names)) {
+    rtn.error = ExceptionError(try_catch, iso, local_ctx);
+    return rtn;
+  }
+
+  uint32_t length = names->Length();
+  const char** strings = (const char**)malloc(length * sizeof(const char*));
+
+  for (uint32_t i = 0; i < length; i++) {
+    Local<Value> name_from_array;
+    if (!names->Get(local_ctx, i).ToLocal(&name_from_array)) {
+      for (i = i - 1; i > 0; i--) {
```

Comment:
Will this miss freeing `strings[0]` if the index doesn't get there? Should the middle param be `i >=0`

Reply:
Good catch!

We can't use `i >=0` though, since the integer is unsigned, so that would always be true.  If we wanted to do that, we could change `i` to int64_t.

Would the code maybe be clearer if it matched the outer loop, e.g.

```c++
      length = i;
      for (i = 0; i < length; i++) {
        free(&strings[i]);
```

or by using a separate loop variable?

Reply:
I think this inner loop is supposed to be from `0 to i` whatever `i` is because it might need to free all strings if an error occurs which means that it might not be equal to `length` and we'd be trying to free elements we don't have yet.

Reply:
Maybe it should be:
```
      for (j = 0; j < i; j --) {
        free(&strings[j]);
      }
      free(strings);
      rtn.error = ExceptionError(try_catch, iso, local_ctx);
      return rtn;
    }
```

Reply:
> I think this inner loop is supposed to be from `0 to i`

right, which is my code snippet started with `length = i;`

Your suggestion is what I meant by "or by using a separate loop variable", except that the for loop line needs to declare the `j` variable by prefixing its assignment with its type: `for (uint32_t j = 0; j < i; j++) {`